### PR TITLE
Fix ConnectionHandlerImpl matching correct any address

### PR DESCRIPTION
**Pull request by soulxu**
Commit Message: Fix ConnectionHandlerImpl matching correct any address
Additional Description:
* Fix https://github.com/envoyproxy/envoy/pull/19362 using the `Network::Utility::getIpv4AnyAddress` with wrong way. `Utility::getIpv4AnyAddress` returns a singleton of any address instance, it won't work with port parameter. This also leads to istio integration test failed https://github.com/istio/istio/pull/37004
* Add more checks for avoid return shutdown listener when redirect.
* Add assertion ensure the redirect won't consider non-ip address.

Risk Level:low
Testing: unittest
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

Signed-off-by: He Jie Xu <hejie.xu@intel.com>
Co-authored-by: Yuchen Dai <silentdai@gmail.com>

### CODE REVIEW

**Comment by soulxu** [comment]
@howardjohn I don't know if there is way to test this fix on the istio side. Hope this can fix the problem. Sorry for making this regression.

cc @lambdai

**Comment by howardjohn** [comment]
I can try it out hopefully sometime today

**Comment by soulxu** [comment]
> I can try it out hopefully sometime today

thanks

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:265]: do you mean here https://github.com/envoyproxy/envoy/pull/19802/files/dfc552724f1883a2c56c3269d46969e78bbcded2#diff-fe538327af200510b78cadc759f2c3af989c822a370d3f8a005b92bc59568aa5L268,

It already has the check.

File: test/server/connection_handler_test.cc
Comment (soulxu) [reply test/server/connection_handler_test.cc]: done

File: test/server/connection_handler_test.cc
Comment (soulxu) [reply test/server/connection_handler_test.cc]: done

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:265]: But I added a test for that branch.

**Comment by soulxu** [comment]
/wait

waiting for istio side test.

File: source/server/connection_handler_impl.cc
Comment (lambdai) [reply source/server/connection_handler_impl.cc:265]: Sorry, I missed that. Thanks!

**Comment by howardjohn** [comment]
I think its still failing...

**Comment by soulxu** [comment]
cc @ggreenway

**Comment by ggreenway** [comment]
@howardjohn can you test this branch and see if it fixes the Istio issues?

**Comment by howardjohn** [comment]
> @howardjohn can you test this branch and see if it fixes the Istio issues?

We don't have the best infrastructure to test arbitrary Envoy builds right now, and the issue was not 100% reproducible (but close), but it appears to be fixed. About 90% confident (where 100% confidence will come after it lands and our CI pulls it in). Thanks for all the work here!

**Review by ggreenway — APPROVED** [review]
LGTM. @lambdai any other comments?

/wait-any

**Comment by soulxu** [comment]
> > @howardjohn can you test this branch and see if it fixes the Istio issues?
> 
> We don't have the best infrastructure to test arbitrary Envoy builds right now, and the issue was not 100% reproducible (but close), but it appears to be fixed. About 90% confident (where 100% confidence will come after it lands and our CI pulls it in). Thanks for all the work here!

Is that possible I compiled an envoy and upload somewhere, then I trigger by a PR on istio side to modify those address
https://github.com/istio/istio/blob/e448c96013adc84de2864c586d9012f9aad7262c/Makefile.core.mk#L130-L134

**Comment by lambdai** [comment]
> LGTM. @lambdai any other comments?
> 
> /wait-any

LGTM. Thanks!

**Comment by lambdai** [comment]
> > > @howardjohn can you test this branch and see if it fixes the Istio issues?
> > 
> > 
> > We don't have the best infrastructure to test arbitrary Envoy builds right now, and the issue was not 100% reproducible (but close), but it appears to be fixed. About 90% confident (where 100% confidence will come after it lands and our CI pulls it in). Thanks for all the work here!
> 
> Is that possible I compiled an envoy and upload somewhere, then I trigger by a PR on istio side to modify those address https://github.com/istio/istio/blob/e448c96013adc84de2864c586d9012f9aad7262c/Makefile.core.mk#L130-L134

the istio/proxy commits could use any commit in envoyproxy/proxy, but only envoy maintainers can create such a commit.

Also, for this particular failure, istio/proxy doesn't detect either.

**Comment by ggreenway** [comment]
@howardjohn please let us know if the istio issue is resolved, after you're able to test with this commit.

**Comment by howardjohn** [comment]
We are good to go :+1: . Thanks everyone

**Comment by soulxu** [comment]
> We are good to go 👍 . Thanks everyone

thanks!

**Pull request merged**
